### PR TITLE
Fixed performance issue with file storage

### PR DESCRIPTION
**Pull request by g4s8**
#171 - using custom IO objects instead of vertx: improved performance of 1Gb file writing about 10x, on local machine it changes from 10-13 seconds to 0.7-0.9 seconds

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#172](https://github.com/artipie/asto/issues/172) is assigned to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/fe37f0aa-87e1-401b-997b-ccf101bf37cb); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/test/java/com/artipie/asto/FileStorageTest.java [hunk 1]
```diff
@@ -139,51 +122,58 @@ void move() {
     @Test
     @EnabledIfSystemProperty(named = "test.storage.file.huge", matches = "true|on")
     void saveAndLoadHugeFiles() throws Exception {
-        // @checkstyle MagicNumberCheck (30 lines)
-        // @checkstyle MethodBodyCommentsCheck (30 lines)
         final Key key = new Key.From("huge");
```
Comment (olenagerasimova) [inline src/test/java/com/artipie/asto/FileStorageTest.java:125]: @g4s8 seems like `key` can be inlined

File: src/test/java/com/artipie/asto/FileStorageTest.java [hunk 2]
```diff
@@ -139,51 +122,58 @@ void move() {
     @Test
     @EnabledIfSystemProperty(named = "test.storage.file.huge", matches = "true|on")
     void saveAndLoadHugeFiles() throws Exception {
```
Comment (olenagerasimova) [inline src/test/java/com/artipie/asto/FileStorageTest.java:124]: @g4s8 this test has no assertions... Maybe we can at least check that storage contains file with given name?

File: src/test/java/com/artipie/asto/FileStorageTest.java [hunk 3]
```diff
@@ -139,51 +122,58 @@ void move() {
...
-        ).get();
-        MatcherAssert.assertThat(
-            hash,
-            new IsEqual<>("13b0e1029eae62b2cde3e918d384ce704319a1eca1cf268b962cb34dbbab04e4")
-        );
+    }
+
+    /**
+     * Provider of byte buffers for write test.
+     * @since 0.1
```
Comment (olenagerasimova) [inline src/test/java/com/artipie/asto/FileStorageTest.java:137]: @g4s8 since 0.2

**Comment by g4s8** [comment]
@olenagerasimova thanks, I added `@Timeout` annotation to fail the test if it doesn't complete in one second and file size assertion

**Comment by olegmoz** [comment]
@g4s8 the big issue I see with this PR is that you are replacing Vert.x file operations with blocking ones from `Files` class. This may result in blocking threads in ForkJoin pool and so in blocking whole application. I'd propose to use separate thread pool for blocking I/O operations on files to prevent whole application degradation when file system is slow.

**Comment by Sammers21** [comment]
@g4s8, how can you proof that the solution you wrote outperforms the existing one?

**Comment by Sammers21** [comment]
@g4s8, who will support the library you wrote **rio** when you will leave the project?

**Comment by g4s8** [comment]
> @g4s8 the big issue I see with this PR is that you are replacing Vert.x file operations with blocking ones from `Files` class. This may result in blocking threads in ForkJoin pool and so in blocking whole application. I'd propose to use separate thread pool for blocking I/O operations on files to prevent whole application degradation when file system is slow.

@olegmoz I didn't find blocking operations in this PR: all read and write operations are performed in thread pool executor (similar to `AsynchronousFileChannel` default executor), see: https://github.com/g4s8/rio/blob/master/src/main/java/wtf/g4s8/rio/file/File.java#L50

**Comment by g4s8** [comment]
> @g4s8, how can you prove that the solution you wrote outperforms the existing one?

@Sammers21 I added `@Timeout` annotation with one second limit to the test with copy of 1Gb file in `FileStorage` test, previous implementation takes about 10 seconds for the same test, you can compare it locally

**Comment by olegmoz** [comment]
> @olegmoz I didn't find blocking operations in this PR: all read and write operations are performed in thread pool executor (similar to `AsynchronousFileChannel` default executor), see: https://github.com/g4s8/rio/blob/master/src/main/java/wtf/g4s8/rio/file/File.java#L50
That's my bad, no blocking ops in default executor indeed.
Looking at `File` code though, I'd suggest to avoid using unlimited `CachedThreadPool` as it will leak threads when file system performance becomes a bottleneck.

**Comment by g4s8** [comment]
@olegmoz thanks, created fixed thread pool with `availableProcessors` size from `Runtime`

**Comment by 0crat** [comment]
Job was finished in 46 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@olegmoz please make sure you start all your comments with the name of the user they are referred to, [see](https://github.com/artipie/asto/pull/172#issuecomment-629171114)

**Comment by sereshqua** [comment]
@olegmoz ping

**Comment by jeffzhu503** [comment]
" improved performance of 1Gb file writing about 10x, on local machine it changes from 10-13 seconds to 0.7-0.9 seconds".  Ideally you want to perf test using a standard machine in the Cloud, and your customers in China and reproduce the same results if they want.  Using a local machine is a good first step before the check-in.  After the check-in, we must establish a performance testing process and measure it in a regular base even without any perf related check-in.  For example, every week, you publish the perf result somewhere to the public.

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Pull request merged**